target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c

```diff
@@ -172,13 +172,10 @@ void start_batch_gemm() {
 void wait_batch_gemm() {
     uint32_t break_poll;
 
-    while (1) {
-        // poll the state CSR[1] to see if GEMM is still busy
-        break_poll = read_csr(0x3cf);
-        if ((break_poll >> 1) == 1) {
-            break;
-        };
-    };
+    // Enable SNAX barrier
+    write_csr(0x7c3, 1);
```

Comment:
@JosseVanDelm Good question, you only need to set it once actually.

Now that I think about it, would you like me to make separate "enable" and "wait" functions?

I also received comments from the last PR that it may not be necessary. But this is a pre-emptive feature for multi-accelerator masking. For example, if a Snitch controls, say 3 accelerators and not all accelerators will be used (probably the 1st and 3rd only) then we can mask the 2nd by disabling the barrier just for the 2nd and just call one trigger: `write_csr(0x7c4, 0);` to activate the entire barrier.

Reply:
I think making an enable or boot function would be nice, but I must say I don't fully understand why this is necessary. If you don't use your 2nd accelerator, it's just in the "ready state" and also doesn't halt the barrier? Why would you still need to disable it? If that makes sense?
If this is not strictly necessary I would remove it, it could lead to some difficult to debug situations? Does that make sense?

Reply:
Yes! These are good points. Wait let me explain, for documentation purposes too, why I made this. It was a conflict of ideas back then but here are a few important features:

* *How does the SNAX barrier work?*
When the SNAX barrier is waiting, it waits for either a pulse or a sticky done signal (sticky means it's stuck to logic high so you could think of it as some sort of busy signal). Once any of these signals arrive then the barrier is released. See figure:

![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/5f3db622-7ef0-4959-9abe-471af0959b11)

* *Why did I have to consider pulse or sticky?*
Sticky seems to be the most convenient. Well, I blame the HWPE accelerator because it's not sticky. It gave a pulse "done" signal only (with the available control signals it has). That's when I thought that I needed to consider a pulse signal instead. I originally wanted a sticky because it does make the most sense and we can get the ignore the enable signal instead. Since a sticky done signal means an accelerator is idle. So that's why I thought of having both.
* *So why have the enable signal?* 
This is useful in the case of multiple accelerators with the assumption that all of them use **pulse** done signals. In the same example of using 3 accelerators. Let's say accelerators 1 and 3 are used. Now, let's say there is **no** enable signal. A sample pseudo-sequence would be:

1. Start accelerators 1 and 3 only -> assuming it takes several cycles
2. Call SNAX barrier -> at this point the barrier waits for the 3 ports
3. Accelerators 1 and 3 send done pulses 
4. SNAX barrier is stuck because pulse from 2nd accelerator never arrives

That's why the **masking** mechanism would be useful. At least in this scenario that's where I see it.

* *Why not restrict everything to just be sticky?* 
Possible! I mean yeah we could require that. However, the HWPE accelerator is not using this mechanism... Can I do some kind of wrapping? Probably yes but that's like *bandage* way through an external wrapper. If a start signal arrives then it pulse the done signal low and the pulse from the original base HWPE sets that same done signal to sticky high.

So yeah it was a conflicting thought and this mechanism was trying to cover both sticky and pulse inputs.

I hope this makes sense? 😆

Reply:
@rgantonio yes that makes a lot of sense.
It's nice to have these options available, but I'd recommend to recomminding sticky then. People who are too far in none-sticky can still do none-sticky if necessary indeed @rgantonio!
We might want to design a stable C API that has some setup routine that can just enable it for the stickies and leaves them be there :)

Reply:
@JosseVanDelm Alrighty!

I will make an issue to make things sticky 😄 THANK YOUU!!!